address-space-controller/src/main/java/io/enmasse/controller/AddressSpaceController.java

```diff
@@ -72,14 +78,55 @@ public void start() throws Exception {
         if (options.isInstallDefaultResources()) {
             configureDefaultResources(controllerClient, options.getResourcesDir());
         }
+        AddressSpaceSchemaUpdater addressSpaceSchemaUpdater = new AddressSpaceSchemaUpdater(controllerClient);
         CachingSchemaProvider schemaProvider = new CachingSchemaProvider();
+        schemaProvider.registerListener(addressSpaceSchemaUpdater);
+
         schemaApi.watchSchema(schemaProvider, options.getResyncInterval());
         Kubernetes kubernetes = new KubernetesHelper(controllerClient.getNamespace(), controllerClient, options.getTemplateDir(), isOpenShift);
 
-        AddressSpaceApi addressSpaceApi = new ConfigMapAddressSpaceApi(controllerClient, options.getVersion());
+        AddressSpaceApi addressSpaceApi = KubeAddressSpaceApi.create(controllerClient, null, options.getVersion());
         EventLogger eventLogger = options.isEnableEventLogger() ? new KubeEventLogger(controllerClient, controllerClient.getNamespace(), Clock.systemUTC(), "address-space-controller")
                 : new LogEventLogger();
 
+        // Convert all address spaces from configmaps to CRD variant
+        ConfigMapAddressSpaceApi legacyAddressSpaceApi = new ConfigMapAddressSpaceApi(controllerClient, options.getVersion());
+        Map<AddressSpace, List<Address>> converted = new HashMap<>();
+
+        for (AddressSpace addressSpace : legacyAddressSpaceApi.listAllAddressSpaces()) {
+            try {
+                addressSpace.getMetadata().setResourceVersion(null);
+                addressSpaceApi.createAddressSpace(addressSpace);
+                AddressApi addressApi = addressSpaceApi.withAddressSpace(addressSpace);
+                AddressApi legacyAddressApi = legacyAddressSpaceApi.withAddressSpace(addressSpace);
+                List<Address> convertedAddresses = new ArrayList<>();
+                for (Address address : legacyAddressApi.listAddresses(addressSpace.getMetadata().getNamespace()).stream().filter(address -> address.getMetadata().getName().startsWith(addressSpace.getMetadata().getName())).collect(Collectors.toList())) {
+                    address.getMetadata().setResourceVersion(null);
+                    addressApi.createAddress(address);
+                    convertedAddresses.add(address);
+                    log.info("Converted address {} in {} from ConfigMap to Address CRD", address.getMetadata().getName(), address.getMetadata().getNamespace());
+                }
+                converted.put(addressSpace, convertedAddresses);
+                log.info("Converted address space {} in {} from ConfigMap to AddressSpace CRD", addressSpace.getMetadata().getName(), addressSpace.getMetadata().getNamespace());
```

Comment:
I wonder if we should emit one kubernetes event for each *addressspace* converted, with the message including a count of the number of addresses converted.  I also think logging an single event corresponding to conversion failure might be useful too.